### PR TITLE
257yk1x make examples working

### PR DESCRIPTION
**Pull request by shubham-shay-te**

### CODE REVIEW

File: examples/objects_core_lib/core_lib/data_layers/data_access/objects_data_access.py [hunk 1]
```diff
@@ -8,18 +9,17 @@
...
+        self.data_session_factory = data_session_factory
 
-    def get_object(self, bucket_name: str, object_name: str):
-        with tempfile.TemporaryFile() as file_name:
-            with self.get_session() as s3:
-                s3.download_fileobj(bucket_name, object_name, file_name)
-                file_content = path(file_name).bytes()
-        return file_content
+    def get_object(self, bucket_name: str, object_name: str, key: str):
+        with self.data_session_factory.get() as s3:
```
Comment (shay-te) [inline examples/objects_core_lib/core_lib/data_layers/data_access/objects_data_access.py:16]: What is the Bytes.io?

File: examples/objects_core_lib/core_lib/data_layers/service/objects_service.py [hunk 2]
```diff
@@ -6,8 +6,8 @@ class ObjectsService(Service):
...
         self.data_access = data_access
 
-    def get_object(self, bucket_name: str, object_name: str):
-        return self.data_access.get_object(bucket_name, object_name)
+    def get_object(self, bucket_name: str, object_name: str, key: str):
+        return self.data_access.get_object(bucket_name, object_name, key)
 
-    def set_object(self, bucket_name: str, value):
-        self.data_access.set_object(bucket_name, value)
+    def set_object(self, bucket_name: str, value, key: str):
```
Comment (shay-te) [inline examples/objects_core_lib/core_lib/data_layers/service/objects_service.py:12]: use the same order as in the `get_object` function
def set_object(self, bucket_name: str,, key: str, value):

File: examples/test_core_lib/core_lib/data_layers/data_access/crud_user_data_access.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+from core_lib.data_layers.data.handler.sql_alchemy_data_handler_registry import SqlAlchemyDataHandlerRegistry
+from core_lib.data_layers.data_access.db.crud.crud import CRUD
+from core_lib.data_layers.data_access.db.crud.crud_soft_data_access import CRUDSoftDataAccess
+from core_lib.helpers.validation import is_email
+from core_lib.rule_validator.rule_validator import ValueRuleValidator, RuleValidator
+from examples.test_core_lib.core_lib.data_layers.data.db.user import User
+
+
+class CrudUserDataAccess(CRUDSoftDataAccess):
+    user_rule_validators = [
```
Comment (shay-te) [inline examples/test_core_lib/core_lib/data_layers/data_access/crud_user_data_access.py:12]: define the rules outside the class. on top of it

File: examples/test_core_lib/core_lib/data_layers/service/user_service.py [hunk 4]
```diff
@@ -1,23 +1,27 @@
+from core_lib.data_layers.data_access.db.crud.crud_soft_data_access import CRUDSoftDataAccess
 from core_lib.data_transform.result_to_dict import ResultToDict
 from core_lib.data_layers.service.service import Service
+from core_lib.error_handling.not_found_decorator import NotFoundErrorHandler
 from examples.test_core_lib.core_lib.data_layers.data_access.user_data_access import UserDataAccess
 
 
 class UserService(Service):
-    def __init__(self, data_access: UserDataAccess):
+    def __init__(self, data_access: UserDataAccess or CRUDSoftDataAccess):
```
Comment (shay-te) [inline examples/test_core_lib/core_lib/data_layers/service/user_service.py:9]: `UserDataAccess or CRUDSoftDataAccess` which one is it?

File: requirements.txt [hunk 5]
```diff
@@ -16,13 +18,19 @@ pymongo
 python-memcached
 cryptography
 boto3
+boto
```
Comment (shay-te) [inline requirements.txt:21]: we have `boto3` why `boto` is also needed?

File: tests/test_combined_example.py [hunk 6]
```diff
@@ -134,3 +148,24 @@ def test_03_core_lib_test_user_create_get(self):
...
+            'updated_at': db_data['created_at'],
+            'created_at': db_data['created_at'],
+            'deleted_at': None,
+        })
+        self.assertDictEqual(db_data, user_data_crud)
+
+        core_lib.test.user_crud.update(user['id'], {'email': 'jon@doe.com'})
+        db_data = core_lib.test.user_crud.get(user['id'])
+        self.assertEqual(db_data['email'], 'jon@doe.com')
+        self.assertNotEqual(db_data['created_at'], db_data['updated_at'])
```
Comment (shay-te) [inline tests/test_combined_example.py:167]: assert greater than is more suitable here

File: tests/test_example_objectscorelib.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+import unittest
+
+import boto3
+from moto import mock_s3
+from omegaconf import OmegaConf
+
+from examples.objects_core_lib.core_lib.objects_core_lib import ObjectsCoreLib
+
+
+class ObjectsCoreLibTesting(ObjectsCoreLib):
```
Comment (shay-te) [inline tests/test_example_objectscorelib.py:11]: what is the purpose of this class?

File: tests/test_example_objectscorelib.py [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+class TestExamples(unittest.TestCase):
+    @mock_s3
+    def test_object_core_lib(self):
+        upload_file_path = os.path.join(os.path.dirname(__file__), 'test_data\koala.jpeg')
+        download_file_path = os.path.join(os.path.dirname(__file__), 'test_data\koalaAWS.jpeg')
+
+        objects_core_lib = ObjectsCoreLibTesting()
+
+        conn = boto3.resource('s3', region_name='us-east-1')
+        conn.create_bucket(Bucket='mybucket')
```
Comment (shay-te) [inline tests/test_example_objectscorelib.py:27]: 1. 'mybucket' is used more than 1 time. therefor use a variable
2. please call it in a name that represents his purpose. `test_bucket` for example!

File: tests/test_example_objectscorelib.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+        objects_core_lib.object.set_object('mybucket', upload_file_path, 'koala.jpeg')
+        obj = objects_core_lib.object.get_object('mybucket', download_file_path, 'koala.jpeg')
+
+        self.assertEqual(download_file_path, obj.name)
```
Comment (shay-te) [inline tests/test_example_objectscorelib.py:32]: the correct test is. 
read the uploaded file content and compare it to the downloaded file content

File: tests/test_example_testcorelib.py [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+        cls.core_lib = sync_create_start_core_lib()
+
+    def test_example_crud(self):
+        user = TestExamples.core_lib.user_crud.create(user_data)
+        db_data = TestExamples.core_lib.user_crud.get(user['id'])
+        self.assertDictEqual(db_data, user)
+
+        TestExamples.core_lib.user_crud.update(user['id'], {'email': 'jon@doe.com'})
+        db_data = TestExamples.core_lib.user_crud.get(user['id'])
+        self.assertEqual(db_data['email'], 'jon@doe.com')
```
Comment (shay-te) [inline tests/test_example_testcorelib.py:52]: if a value is used more than once please use a variable!

File: tests/test_example_testcorelib.py [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+        self.assertNotEqual(db_data['created_at'], db_data['updated_at'])
+
+        TestExamples.core_lib.user_crud.delete(user['id'])
```
Comment (shay-te) [inline tests/test_example_testcorelib.py:55]: User.id.key?

File: tests/test_example_testcorelib.py (hunk 10 above)
Comment (shay-te) [inline tests/test_example_testcorelib.py:52]: User.email.key?

File: tests/test_example_testcorelib.py [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
+    def test_example_user_data_access(self):
+        user = TestExamples.core_lib.user.create(user_data)
+        db_data = TestExamples.core_lib.user.get(user['id'])
+        self.assertDictEqual(db_data, user)
+
+        TestExamples.core_lib.user.update(user['id'], {'gender': User.Gender.FEMALE.value})
+        db_data = TestExamples.core_lib.user.get(user['id'])
+        self.assertEqual(db_data['gender'], User.Gender.FEMALE.value)
+        self.assertNotEqual(db_data['created_at'], db_data['updated_at'])
```
Comment (shay-te) [inline tests/test_example_testcorelib.py:67]: User.created_at.key?
User.updated_at.key?

**Review by shay-te — CHANGES_REQUESTED** [review]

File: examples/test_core_lib/core_lib/data_layers/service/user_service.py (hunk 4 above)
Comment (shubham-shay-te) [reply examples/test_core_lib/core_lib/data_layers/service/user_service.py:9]: if you see in the `examples/test_core_lib/core_lib/test_core_lib.py` file we have 2 `UserService` instances one is passing `UserDataAccess` whereas the other is passing `CrudUserDataAccess`, to handle this the following change was made

File: requirements.txt (hunk 5 above)
Comment (shubham-shay-te) [reply requirements.txt:21]: there's an import for that `from boto import Config` in `object_core_lib` which I have removed as the config was causing an error while mocking, do we need to keep it? and how do we configure it with corelib

File: tests/test_example_objectscorelib.py (hunk 7 above)
Comment (shubham-shay-te) [reply tests/test_example_objectscorelib.py:11]: for passing config to `object_core_lib` for now the current config will require docker for env variables, I tried to edit the config but boto3 fails at that time, so I made this class to pass a basic config file to `object_core_lib`

File: examples/test_core_lib/core_lib/test_core_lib.py [hunk 13]
```diff
@@ -29,4 +30,6 @@ def __init__(self):
 
         self.test = Test()
         self.user = UserService(UserDataAccess(db_data_session))
+        self.user_crud = UserService(CrudUserDataAccess(db_data_session))
```
Comment (shay-te) [inline examples/test_core_lib/core_lib/test_core_lib.py:33]: 1. rename CrudUserDataAccess to CustomerDataAccess
2. create CustomerService.

File: examples/test_core_lib/core_lib/data_layers/service/user_service.py (hunk 4 above)
Comment (shay-te) [reply examples/test_core_lib/core_lib/data_layers/service/user_service.py:9]: 1. use only UserDataAccess
2. create CustomerService that will accept CustomerDataAccess(CRUD)

File: tests/test_example_objectscorelib.py (hunk 7 above)
Comment (shay-te) [reply tests/test_example_objectscorelib.py:11]: please use ObjectsCoreLib directly

**Review by shay-te — CHANGES_REQUESTED** [review]

File: examples/test_core_lib/core_lib/data_layers/service/customer_service.py [hunk 14]
```diff
@@ -0,0 +1,25 @@
+from core_lib.data_layers.data_access.db.crud.crud_soft_data_access import CRUDSoftDeleteDataAccess
+from core_lib.data_transform.result_to_dict import ResultToDict
+from core_lib.data_layers.service.service import Service
+from core_lib.error_handling.not_found_decorator import NotFoundErrorHandler
+
+
+class CustomerService(Service):
+    def __init__(self, data_access: CRUDSoftDeleteDataAccess):
```
Comment (shay-te) [inline examples/test_core_lib/core_lib/data_layers/service/customer_service.py:8]: the customer data access

File: requirements.txt [hunk 15]
```diff
@@ -19,10 +21,15 @@ boto3
...
 django
 flask
+requests
 
 
 # Test
 python-dotenv
+moto
+freezegun
+setuptools
```
Comment (shay-te) [inline requirements.txt:35]: move to line 23

File: tests/test_combined_example.py [hunk 16]
```diff
@@ -32,6 +32,19 @@
 
 core_lib = CombineCoreLib(config.core_lib)
```
Comment (shay-te) [inline tests/test_combined_example.py:33]: rename `core_lib` to `combine_core_lib`

File: tests/test_example_objectscorelib.py [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+from moto import mock_s3
+
+from examples.objects_core_lib.core_lib.objects_core_lib import ObjectsCoreLib
+
+
+@mock_s3
+class TestExamples(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        config_file = 'config.yaml'
```
Comment (shay-te) [inline tests/test_example_objectscorelib.py:15]: make test util will this function 
```
def load_config():
    if not OblInstange.config:
        path = os.path.join(os.path.dirname(__file__), '..', 'data')
        load_dotenv(dotenv_path=os.path.join(path, '.env'))

        GlobalHydra.instance().clear()
        hydra.initialize(config_path=os.path.join('..', 'data', 'config'), caller_stack_depth=1)
        OblInstange.config = hydra.compose('config.yaml')
    return OblInstange.config
```

File: tests/test_example_objectscorelib.py [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+        hydra.core.global_hydra.GlobalHydra.instance().clear()
+        hydra.initialize(config_path='../examples/objects_core_lib/config')
+        config = hydra.compose(config_file)
+
+        cls.objects_core_lib = ObjectsCoreLib(config.core_lib)
```
Comment (shay-te) [inline tests/test_example_objectscorelib.py:20]: just pass the config

File: tests/test_example_testcorelib.py [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+
+def sync_create_start_core_lib() -> TestCoreLib:
+    [CoreLib.cache_registry.unregister(key) for key in CoreLib.cache_registry.registered()]
+    [CoreLib.observer_registry.unregister(key) for key in CoreLib.observer_registry.registered()]
+    config_file = 'config.yaml'
+    hydra.core.global_hydra.GlobalHydra.instance().clear()
+    hydra.initialize(config_path='../examples/test_core_lib/config')
+    config = hydra.compose(config_file)
+
+    core_lib = TestCoreLib(config.core_lib)
```
Comment (shay-te) [inline tests/test_example_testcorelib.py:34]: this line in def setUp

File: tests/test_example_testcorelib.py [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+    'first_name': 'Jon',
+    'middle_name': 'P',
+    'last_name': 'Doe',
+    'birthday': dattime,
+    'gender': User.Gender.MALE.value,
+}
+
+
+def sync_create_start_core_lib() -> TestCoreLib:
+    [CoreLib.cache_registry.unregister(key) for key in CoreLib.cache_registry.registered()]
```
Comment (shay-te) [inline tests/test_example_testcorelib.py:27]: this on the utils

**Review by shay-te — CHANGES_REQUESTED** [review]

**Review by shay-te — APPROVED** [review]

**Pull request merged**